### DISCUSSION TITLE
Next.js Middleware: Express.js Simplicity vs. Next.js Complexity? Seeking Best Practices

### DISCUSSION THREAD

**Discussion by rayvego** [vercel/next.js#67746] [Help]
### Summary

Hi everyone, I'm transitioning from MERN stack development (Express.js) to Next.js. In Express.js, defining and applying middleware was incredibly intuitive. I'd have separate middleware functions (e.g., authentication, authorization) and simply chain them in the desired order for specific routes:

```javascript
// Express.js example
router.get("/:id/edit", isLoggedIn, isAuthor, catchAsync(campgrounds.renderEditForm));
```

However, Next.js middleware seems quite different. From what I've gathered, all middleware goes into a single middleware.ts file, requiring conditional checks for routes and methods. To chain middleware, recursion seems necessary. For large projects, this approach feels cumbersome and less maintainable.

Am I missing something? Is there a cleaner way to organize and apply Next.js middleware that aligns more closely with the Express.js approach? I'm hoping to find best practices or patterns that simplify middleware management in Next.js.

Thanks in advance for any insights or advice!

**Comment by icyJoseph** [comment]
A Next.js middleware, is not the same middleware, in terms of concept as the Express middleware. 

The middleware in Next.js, is primarily designed, to run in a separate context, that is, it can run in a different location, than the server. 

I think, a more proper way to understand Next.js middleware, is this model: 

- https://docs.aws.amazon.com/lambda/latest/dg/lambda-edge.html

> Note, that this is for didactical purposes. 

Next.js middleware, would be a `viewer request` modifier. 

I think, you might want to take a look at this post: https://github.com/vercel/next.js/discussions/63775
